packages/liveblocks-core/src/connection.ts

```diff
@@ -0,0 +1,756 @@
+                "Connection to WebSocket closed permanently. Won't retry."
+              ),
+          }; // Should not retry, give up
+        }
+
+        // If this is a custom Liveblocks server close reason, back off more
+        // aggressively, and emit a Liveblocks error event...
+        if (e.event.code >= 4000 && e.event.code <= 4100) {
+          return {
+            target: "@connecting.busy",
+            assign: increaseBackoffDelayAggressively,
+            effect: (_, { event }) => {
+              if (event.code >= 4000 && event.code <= 4100) {
+                const err = new LiveblocksError(event.reason, event.code);
+                onLiveblocksError.notify(err);
+              }
+            },
+          };
+        }
+
+        // Consider this close event a temporary hiccup, and try re-opening
+        // a new socket
+        return {
+          target: "@connecting.busy",
+          assign: increaseBackoffDelay,
+        };
+      },
+    });
+
+  // Lastly, register an event handler to listen for window-focus events as
+  // soon as the machine starts, and use it to send itself "WINDOW_GOT_FOCUS"
+  // events.
+  if (typeof document !== "undefined") {
+    const doc = typeof document !== "undefined" ? document : undefined;
+    const win = typeof window !== "undefined" ? window : undefined;
+    const root = win ?? doc;
+
+    fsm.onEnter("*", (ctx) => {
+      function onBackOnline() {
+        fsm.send({ type: "NAVIGATOR_ONLINE" });
+      }
+
+      function onVisibilityChange() {
+        if (doc?.visibilityState === "visible") {
+          fsm.send({ type: "WINDOW_GOT_FOCUS" });
+        }
+      }
+
+      win?.addEventListener("online", onBackOnline);
+      root?.addEventListener("visibilitychange", onVisibilityChange);
+      return () => {
+        root?.removeEventListener("visibilitychange", onVisibilityChange);
+        win?.removeEventListener("online", onBackOnline);
+
+        // Also tear down the old socket when stopping the machine, if there is one
+        teardownSocket(ctx.socket);
+      };
+    });
+  }
+
+  const { statusDidChange, didConnect, didDisconnect } =
+    defineConnectivityEvents(fsm);
+
+  // Install debug logging
+  const cleanup = enableTracing(fsm); // TODO Remove logging in production
+
+  // Start the machine
+  fsm.start();
+
+  return {
+    fsm,
+    cleanup,
+
+    // Observable events that will be emitted by this machine
+    events: {
+      statusDidChange,
+      didConnect,
+      didDisconnect,
+      onMessage: onMessage.observable,
+      onLiveblocksError: onLiveblocksError.observable,
+    },
+  };
+}
+
+/**
+ * The ManagedSocket will set up a WebSocket connection to a room, and maintain
+ * that connection over time.
+ *
+ * It's a light wrapper around the actual FSM that implements the logic,
+ * exposing just a few safe actions and events that can be called or observed
+ * from the outside.
+ */
+export class ManagedSocket<T extends BaseAuthResult> {
```

Comment:
As a dev, I feel like maybe these should be in separate files. One file is my state machine definition, the other is the simple wrapper class.

Reply:
This started as a single class, that defined the state machine inline, but I pulled out the helper function because it was getting so large and the indentation in the class made things super hard to read. Basically, the entire [`createConnectionStateMachine()`](https://github.com/liveblocks/liveblocks/blob/c44a469f6f1045b615398780765a1c15ea9e8de1/packages/liveblocks-core/src/connection.ts#L259-L639) helper function is supposed to ["just" be the constructor](https://github.com/liveblocks/liveblocks/blob/c44a469f6f1045b615398780765a1c15ea9e8de1/packages/liveblocks-core/src/connection.ts#L677-L681)—they're not really different things. For that reason, I'd prefer to keep them in the same file, as they are just one conceptual thing and need to always change together.